fdbserver/ClusterController.actor.cpp

```diff
@@ -3485,6 +3730,58 @@ ACTOR Future<Void> dbInfoUpdater(ClusterControllerData* self) {
 	}
 }
 
+// The actor that periodically monitors the health of tracked workers.
+ACTOR Future<Void> workerHealthMonitor(ClusterControllerData* self) {
+	loop {
+		try {
+			while (!self->goodRecruitmentTime.isReady()) {
+				wait(delay(SERVER_KNOBS->CC_WORKER_HEALTH_CHECKING_INTERVAL));
+			}
+
+			self->degradedServers = self->getServersWithDegradedLink();
+
+			// Compare `self->degradedServers` with `self->excludedDegradedServers` and remove those that have
+			// recovered.
+			for (auto it = self->excludedDegradedServers.begin(); it != self->excludedDegradedServers.end();) {
+				if (self->degradedServers.find(*it) == self->degradedServers.end()) {
+					self->excludedDegradedServers.erase(it++);
+				} else {
+					++it;
+				}
+			}
+
+			if (!self->degradedServers.empty()) {
+				std::string degradedServerString;
+				for (const auto& server : self->degradedServers) {
+					degradedServerString += server.toString() + " ";
+				}
+				TraceEvent("ClusterControllerHealthMonitor").detail("DegradedServers", degradedServerString);
+
+				// Check if the cluster controller should trigger a recovery to exclude any degraded servers from the
+				// transaction system.
+				if (self->shouldTriggerRecoveryDueToDegradedServers()) {
+					if (SERVER_KNOBS->CC_HEALTH_TRIGGER_RECOVERY) {
+						if (self->recentRecoveryCountDueToHealth() < SERVER_KNOBS->CC_MAX_HEALTH_RECOVERY_COUNT) {
+							self->recentHealthTriggeredRecoveryTime.push(now());
+							self->excludedDegradedServers = self->degradedServers;
+							TraceEvent("DegradedServerDetectedAndTriggerRecovery")
+							    .detail("RecentRecoveryCountDueToHealth", self->recentRecoveryCountDueToHealth());
+							self->db.forceMasterFailure.trigger();
+						}
+					} else {
+						self->excludedDegradedServers.clear();
+						TraceEvent("DegradedServerDetectedAndSuggestRecovery");
```

Comment:
can we make the event more straightforward to SRE on which server to exclude before trigger recovery? 
e.g., add a field. 
.detail("Hint", "Exclude based on ClusterControllerHealthMonitor event and manually trigger recovery")